File: foolish-core-scala/src/main/scala/org/foolish/fvm/scubc/FIR.scala
package org.foolish.fvm.scubc

import org.foolish.ast.AST
import scala.jdk.CollectionConverters.*

/**
 * Foolish Internal Representation (FIR).
 * The FIR is the internal representation of computation that holds an AST
 * and tracks evaluation progress.
 */
abstract class FIR(val ast: AST, val comment: Option[String] = None):
  private var initialized: Boolean = false
  private var nyes: Nyes = Nyes.UNINITIALIZED
  private var parentFir: FIR = null  // The FIR that contains this FIR (set during enqueue)

  /**
   * Returns the current NYE state of this FIR.
   */
  def getNyes: Nyes = nyes

  /**
   * Sets the NYE state of this FIR.
   * All changes to a Firoe's Nyes must be made through this method.
   */
  protected def setNyes(newNyes: Nyes): Unit =
    nyes = newNyes

  /**
   * Sets the parent FIR that contains this FIR.
   * This is called when this FIR is enqueued into a parent's braneMind.
   */
  protected[scubc] def setParentFir(parent: FIR): Unit =
    parentFir = parent

  /**
   * Gets the parent FIR that contains this FIR.
   * Returns null if this FIR has no parent (e.g., root brane).
   */
  protected[scubc] def getParentFir: FIR = parentFir

  final def atConstant: Boolean = nyes == Nyes.CONSTANT
  final def atConstanic: Boolean = nyes == Nyes.CONSTANIC

  /**
   * Returns true when nyes >= CONSTANIC (i.e., CONSTANIC OR CONSTANT)
   * This is the Scala equivalent of Java's isConstanic() method.
   */
  def isConstanic: Boolean = nyes == Nyes.CONSTANIC || nyes == Nyes.CONSTANT

  /**
   * Returns true when nyes >= CONSTANT (i.e., CONSTANT only)
   * This is the Scala equivalent of Java's isConstant() method.
   */
  final def isConstant: Boolean = nyes == Nyes.CONSTANT

  /**
   * Checks if this FIR has been initialized.
   */
  protected def isInitialized: Boolean = initialized

  /**
   * Marks this FIR as initialized.
   */
  protected def setInitialized(): Unit =
    initialized = true

  /**
   * Initializes this FIR. Called once before first step.
   * Override in subclasses to add initialization logic.
   */
  protected def initialize(): Unit =
    if !initialized then
      setInitialized()

  /** Perform one step of evaluation on this FIR */
  def step(): Int

  /**
   * Query method returning false if an additional step on this FIR does not change it.
   * Returns true when an additional step would change the FIR.
   * Not Yet Evaluated (NYE) indicates the FIR requires further evaluation steps.
   */
  def isNye: Boolean

  /**
   * Query method returning false only when all identifiers are bound.
   * Returns true if there are unbound identifiers (abstract state).
   */
  def isAbstract: Boolean

  /**
   * Gets the value from this FIR if it represents a simple value.
   * Throws UnsupportedOperationException if not supported.
   */
  def getValue: Long =
    throw UnsupportedOperationException(s"getValue not supported for ${getClass.getSimpleName}")

  /**
   * Returns the "valuable self" of this FIR, resolving wrappers to their significant values.
   * <p>
   * This matches Java's valuableSelf() method behavior:
   * - Returns Some(this) by default (literal constants, branes, etc).
   * - Returns result.valuableSelf() for AssignmentFiroe.
   * - Returns value.valuableSelf() for IdentifierFiroe.
   * - Returns None if the FIR is constanic/unresolved.
   * - Returns null if the FIR shouldn't have been called yet (pre-PRIMED).
   */
  def valuableSelf(): java.util.Optional[FIR] =
    java.util.Optional.of(this)

  /**
   * Gets the BraneFiroe that contains this FIR in its statement list.
   * Chains through parent FIRs until finding one whose parent is a BraneFiroe.
   * <p>
   * The containing brane is the closest BraneFiroe in the FIR hierarchy,
   * representing the brane where this FIR appears as a statement.
   * <p>
   * Parallel expressions (such as operands in a+b) are at the "same height"
   * and all statements in a brane are parallel/same height. The height does
   * NOT deepen with deepening FIR structures - height only changes when
   * crossing brane boundaries.
   *
   * @return the containing BraneFiroe, or null if this FIR is not contained
   *         in a brane (e.g., at root level)
   */
  def getMyBrane: BraneFiroe =
    parentFir match
      case bf: BraneFiroe => bf
      case null => null
      case _ => parentFir.getMyBrane

  /**
   * Gets the index of this FIR in its containing brane's memory.
   * First finds the containing BraneFiroe using getMyBrane(), then returns its index.
   * If not directly in the brane, delegates to parent.
   * <p>
   * The brane index defines the "order of expressions" within a height level.
   * All statements at the same brane level are parallel/same height, and the
   * index orders them for operations like unanchored backward search.
   * <p>
   * Note: The index is for the statement containing this FIR, not necessarily
   * this exact FIR object. For example, if this is a sub-expression of an
   * assignment, it returns the assignment's index in the brane.
   *
   * @return the index in the containing brane's memory (0-based), or -1 if
   *         this FIR is not in a brane (root level)
   */
  def getMyBraneIndex: Int =
    val containingBrane = getMyBrane
    if containingBrane == null then
      -1
    else
      val directIndex = containingBrane.getIndexOf(this)
      if directIndex != -1 then
        directIndex
      else
        // Not directly in the brane - delegate to parent
        if parentFir != null then
          parentFir.getMyBraneIndex
        else
          -1

  /**
   * Gets the FiroeWithBraneMind container (BraneFiroe or ConcatenationFiroe)
   * that contains this FIR in its braneMemory.
   * Chains through parent FIRs to find the container.
   * <p>
   * This method is used by unanchored seeks to find the scope of their backward search.
   * The search scope includes both branes and concatenations (joined branes).
   *
   * Note: This method only returns BraneFiroe or ConcatenationFiroe, not other
   * FiroeWithBraneMind subclasses like AssignmentFiroe or BinaryFiroe.
   *
   * @return the containing BraneFiroe or ConcatenationFiroe, or null if not contained
   */
  def getMyBraneContainer: BraneFiroe | ConcatenationFiroe =
    parentFir match
      case bf: BraneFiroe => bf
      case cf: ConcatenationFiroe => cf
      case null => null
      case _ => parentFir.getMyBraneContainer

  /**
   * Gets the index of this FIR in its containing brane container's memory.
   * Chains through parent FIRs to find the statement-level FIR, then returns its position.
   * <p>
   * Works with both BraneFiroe and ConcatenationFiroe containers.
   *
   * Note: This method only considers BraneFiroe or ConcatenationFiroe as containers,
   * not other FiroeWithBraneMind subclasses like AssignmentFiroe or BinaryFiroe.
   *
   * @return the index in the containing memory (0-based), or -1 if not in a container
   */
  def getMyBraneContainerIndex: Int =
    parentFir match
      case null =>
        System.out.println(s"DEBUG getMyBraneContainerIndex: null parentFir returning -1 for $this")
        -1
      case bf: BraneFiroe =>
        val idx = bf.getIndexOf(this)
        System.out.println(s"DEBUG getMyBraneContainerIndex: parent is BraneFiroe (hashCode=${System.identityHashCode(bf)}), getIndexOf(this)=$idx for $this (hashCode=${System.identityHashCode(this)})")
        if idx < 0 then
          System.out.println(s"DEBUG getMyBraneContainerIndex: WARNING - idx < 0, searching parent chain")
        idx
      case cf: ConcatenationFiroe =>
        val idx = cf.getIndexOf(this)
        System.out.println(s"DEBUG getMyBraneContainerIndex: parent is ConcatenationFiroe (hashCode=${System.identityHashCode(cf)}), getIndexOf(this)=$idx for $this (hashCode=${System.identityHashCode(this)})")
        if idx < 0 then
          System.out.println(s"DEBUG getMyBraneContainerIndex: WARNING - idx < 0, searching parent chain")
        idx
      case _ =>
        val parentIdx = parentFir.getMyBraneContainerIndex
        System.out.println(s"DEBUG getMyBraneContainerIndex: parent is ${parentFir.getClass.getSimpleName} (hashCode=${System.identityHashCode(parentFir)}), chaining to parent, parentIdx=$parentIdx for $this (hashCode=${System.identityHashCode(this)})")
        parentIdx

object FIR:
  /** Creates a FIR from an AST expression */
  def createFiroeFromExpr(expr: AST.Expr): FIR = expr match
    case literal: AST.IntegerLiteral => ValueFiroe(literal, literal.value())
    case binary: AST.BinaryExpr => BinaryFiroe(binary)
    case unary: AST.UnaryExpr => UnaryFiroe(unary)
    case ifExpr: AST.IfExpr => IfFiroe(ifExpr)
    case brane: AST.Brane => BraneFiroe(brane)
    case assignment: AST.Assignment => AssignmentFiroe(assignment)
    case identifier: AST.Identifier => IdentifierFiroe(identifier)
    case regexpSearch: AST.RegexpSearchExpr =>
      if DerefSearchFiroe.isExactMatch(regexpSearch.pattern()) then
        new DerefSearchFiroe(regexpSearch)
      else
        new RegexpSearchFiroe(regexpSearch)
    case oneShotSearch: AST.OneShotSearchExpr => new OneShotSearchFiroe(oneShotSearch)
    case dereferenceExpr: AST.DereferenceExpr =>
      val synthetic = new AST.RegexpSearchExpr(dereferenceExpr.anchor(), org.foolish.ast.SearchOperator.REGEXP_LOCAL, dereferenceExpr.coordinate().toString)
      new DerefSearchFiroe(synthetic, dereferenceExpr)
    case seekExpr: AST.SeekExpr => SeekFiroe(seekExpr)
    case unanchoredSeekExpr: AST.UnanchoredSeekExpr => UnanchoredSeekFiroe(unanchoredSeekExpr)
    case concatenation: AST.Concatenation => ConcatenationFiroe(concatenation)
    case _ => ValueFiroe(null, 0L) // Placeholder for unsupported types

  /**
   * Unwraps a Constanicable FIR to its resolved value.
   *
   * Follows the wrapper chain while the FIR is constanic:
   * - IdentifierFiroe -> its value
   * - AssignmentFiroe -> its result
   * - SearchFiroe -> its searchResult
   *
   * Returns the first FIR in the chain that is either:
   * - Not a Constanicable
   * - A Constanicable with null or self-referential result
   */
  def unwrapConstanicable(fir: FIR): FIR =
    var current = fir
    while current != null do
      if current.isInstanceOf[Constanicable] then
        val result = current.asInstanceOf[Constanicable].getResult
        if result == null || result == current then return current
        current = result
      else
        return current
    current

File: foolish-core-java/src/main/java/org/foolish/fvm/ubc/FIR.java
package org.foolish.fvm.ubc;

import org.foolish.ast.AST;
import org.foolish.ast.SearchOperator;

/**
 * Foolish Internal Representation (FIR).
 * <p>
 * Base class for all FIRs. Holds an AST node and tracks evaluation progress via Nyes state.
 * <p>
 * See {@code projects/FIR-Invariances.md} for constraints, state machine, and method contracts.
 */
public abstract class FIR implements Cloneable {
    protected final AST ast;
    protected final String comment;
    private boolean initialized;
    protected Nyes nyes;
    private FIR parentFir = null;
    private final boolean ai;

    protected FIR(AST ast, String comment, boolean ai) {
        this.ast = ast;
        this.comment = comment;
        this.initialized = false;
        this.nyes = Nyes.UNINITIALIZED;
        this.ai = ai;
    }

    protected FIR(AST ast, String comment) {
        this(ast, comment, false);
    }

    protected FIR(AST ast) {
        this(ast, null, false);
    }

    protected FIR(String comment) {
        this(null, comment, true);
    }

    public AST ast() {
        return ast;
    }

    public String comment() {
        return comment;
    }

    public boolean isAi() {
        return ai;
    }

    public AST.SourceLocation sourceLocation() {
        return ast != null ? ast.sourceLocation() : AST.SourceLocation.UNKNOWN;
    }

    public AST.SourceLocation nearestSourceLocation() {
        AST.SourceLocation loc = sourceLocation();
        if (loc != AST.SourceLocation.UNKNOWN) {
            return loc;
        }
        FIR current = parentFir;
        while (current != null) {
            loc = current.sourceLocation();
            if (loc != AST.SourceLocation.UNKNOWN) {
                return loc;
            }
            current = current.parentFir;
        }
        return AST.SourceLocation.UNKNOWN;
    }

    public int getMyBraneStatementNumber() {
        BraneFiroe containingBrane = getMyBrane();
        if (containingBrane != null) {
            int directIndex = containingBrane.getStatementIndex(this);
            if (directIndex != -1) {
                return directIndex;
            }
            if (parentFir != null) {
                return parentFir.getMyBraneStatementNumber();
            }
        }
        return -1;
    }

    public String getLocationDescription() {
        StringBuilder sb = new StringBuilder();
        sb.append(nearestSourceLocation());
        int stmtNum = getMyBraneStatementNumber();
        if (stmtNum >= 0) {
            sb.append(" (statement #").append(stmtNum + 1).append(")");
        }
        if (isAi()) {
            sb.append(" [auto-generated]");
        }
        return sb.toString();
    }

    public String formatErrorMessage(String message) {
        StringBuilder sb = new StringBuilder();
        ExecutionContext ctx = ExecutionContext.getCurrent();
        sb.append(ctx != null ? ctx.getSourceFilename() : "unknown.foo").append(":");
        AST.SourceLocation loc = nearestSourceLocation();
        if (loc != AST.SourceLocation.UNKNOWN) {
            sb.append("line ").append(loc.line());
        } else if (isAi()) {
            sb.append("[ai]");
        } else {
            sb.append("unknown");
        }
        sb.append(" Brane@").append(getMyBraneStatementNumber()).append(" - ").append(message);
        return sb.toString();
    }

    protected boolean isInitialized() {
        return initialized;
    }

    protected void setInitialized() {
        this.initialized = true;
    }

    protected void markInitialized() {
        this.initialized = true;
        setNyes(Nyes.INITIALIZED);
    }

    protected Nyes getNyes() {
        return nyes;
    }

    protected void setNyes(Nyes nyes) {
        if (this.nyes == null) {
            this.nyes = nyes;
            return;
        }
        if (this.nyes == Nyes.CONSTANIC) {
            if (nyes != Nyes.CONSTANT && nyes != Nyes.CONSTANIC) {
                throw new IllegalStateException(formatErrorMessage(
                    "Cannot change state from CONSTANIC to " + nyes + " - FIR is immutable once CONSTANIC"));
            }
            if (nyes == Nyes.CONSTANT) {
                this.nyes = nyes;
            }
        } else if (this.nyes == Nyes.CONSTANT) {
            if (nyes != Nyes.CONSTANT) {
                throw new IllegalStateException(formatErrorMessage(
                    "Cannot change state from CONSTANT to " + nyes + " - FIR is immutable once CONSTANT"));
            }
        } else if (nyes.ordinal() < this.nyes.ordinal()) {
            throw new IllegalStateException(formatErrorMessage(
                "Cannot transition backward from " + this.nyes + " to " + nyes));
        } else {
            this.nyes = nyes;
        }
    }

    protected void setParentFir(FIR parent) {
        this.parentFir = parent;
    }

    protected FIR getParentFir() {
        return parentFir;
    }

    public boolean atConstant() {
        return getNyes() == Nyes.CONSTANT;
    }

    public boolean atConstanic() {
        return getNyes() == Nyes.CONSTANIC;
    }

    /**
     * Perform one step of evaluation.
     * @return 0 for no-op, 1 for meaningful work
     * @see "projects/FIR-Invariances.md#Step Method Contract"
     */
    public abstract int step();

    public boolean isNye() {
        return getNyes().ordinal() < Nyes.CONSTANIC.ordinal();
    }

    public boolean isConstanic() {
        return getNyes().ordinal() >= Nyes.CONSTANIC.ordinal();
    }

    public boolean isConstant() {
        return getNyes().ordinal() >= Nyes.CONSTANT.ordinal();
    }

    public long getValue() {
        throw new UnsupportedOperationException("getValue not supported for " + getClass().getSimpleName());
    }

    public BraneFiroe getMyBrane() {
        if (parentFir instanceof BraneFiroe bf) {
            return bf;
        }
        return parentFir != null ? parentFir.getMyBrane() : null;
    }

    public FiroeWithBraneMind getMyBraneContainer() {
        if (parentFir instanceof BraneFiroe || parentFir instanceof ConcatenationFiroe) {
            return (FiroeWithBraneMind) parentFir;
        }
        return parentFir != null ? parentFir.getMyBraneContainer() : null;
    }

    public int getMyBraneContainerIndex() {
        if (parentFir instanceof BraneFiroe || parentFir instanceof ConcatenationFiroe) {
            return ((FiroeWithBraneMind) parentFir).getIndexOf(this);
        }
        return parentFir != null ? parentFir.getMyBraneContainerIndex() : -1;
    }

    public static FIR unwrapConstanicable(FIR fir) {
        FIR current = fir;
        while (current != null && current instanceof Constanicable constanicable) {
            FIR result = constanicable.getResult();
            if (result == null || result == current) {
                break;
            }
            current = result;
        }
        return current;
    }

    public FoolishIndex getMyIndex() {
        FoolishIndexBuilder builder = new FoolishIndexBuilder();
        FIR current = this;
        while (current != null) {
            int index = current.getMyBraneStatementNumber();
            if (index == -1) {
                builder.prepend(0);
                break;
            }
            builder.prepend(index);
            current = current.getMyBrane();
        }
        return builder.build();
    }

    public FIR copy(Nyes targetNyes) {
        if (targetNyes != null && nyes != targetNyes) {
            FIR fresh = this.clone();
            fresh.nyes = targetNyes;
            return fresh;
        }
        if (isConstanic()) {
            return this;
        }
        if (this.ast() instanceof AST.Expr expr) {
            return new CMFir(this.ast(), this);
        }
        return this.clone();
    }

    protected static final ThreadLocal<Integer> RECURSION_DEPTH = ThreadLocal.withInitial(() -> 0);

    public java.util.Optional<FIR> valuableSelf() {
        return java.util.Optional.of(this);
    }

    @Override
    protected FIR clone() {
        try {
            return (FIR) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new RuntimeException("Clone not supported for " + getClass().getSimpleName(), e);
        }
    }

    protected FIR cloneConstanic(java.util.Optional<Nyes> targetNyes) {
        return cloneConstanic(this.getParentFir(), targetNyes);
    }

    /**
     * Clones a CONSTANIC/CONSTANT FIR with updated parent chain.
     * @see "projects/FIR-Invariances.md#cloneConstanic Contract"
     */
    protected FIR cloneConstanic(FIR newParent, java.util.Optional<Nyes> targetNyes) {
        if (!isConstanic()) {
            throw new IllegalStateException(formatErrorMessage(
                "cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, but this FIR is in state: " + getNyes()));
        }
        if (isConstant()) {
            return this;
        }
        FIR copy = this.clone();
        copy.setParentFir(newParent);
        copy.nyes = targetNyes.orElse(this.nyes);
        return copy;
    }

    protected static FIR createFiroeFromExpr(AST.Expr expr) {
        return switch (expr) {
            case AST.IntegerLiteral literal -> new ValueFiroe(expr, literal.value());
            case AST.BinaryExpr binary -> new BinaryFiroe(binary);
            case AST.UnaryExpr unary -> new UnaryFiroe(unary);
            case AST.IfExpr ifExpr -> new IfFiroe(ifExpr);
            case AST.Brane brane -> new BraneFiroe(brane);
            case AST.Concatenation concatenation -> new ConcatenationFiroe(concatenation);
            case AST.DetachmentBrane detachBrane -> new DetachmentBraneFiroe(detachBrane);
            case AST.Assignment assignment -> new AssignmentFiroe(assignment);
            case AST.Identifier identifier -> new IdentifierFiroe(identifier);
            case AST.RegexpSearchExpr regexpSearch ->
                DerefSearchFiroe.isExactMatch(regexpSearch.pattern())
                    ? new DerefSearchFiroe(regexpSearch)
                    : new RegexpSearchFiroe(regexpSearch);
            case AST.OneShotSearchExpr oneShotSearch -> new OneShotSearchFiroe(oneShotSearch);
            case AST.DereferenceExpr dereferenceExpr -> {
                AST.RegexpSearchExpr synthetic = new AST.RegexpSearchExpr(
                    dereferenceExpr.anchor(), SearchOperator.REGEXP_LOCAL, dereferenceExpr.coordinate().toString());
                yield new DerefSearchFiroe(synthetic, dereferenceExpr);
            }
            case AST.SeekExpr seekExpr -> new SeekFiroe(seekExpr);
            case AST.UnanchoredSeekExpr unanchoredSeekExpr -> new UnanchoredSeekFiroe(unanchoredSeekExpr);
            case AST.StayFoolishExpr stayFoolish -> {
                FIR innerFir = createFiroeFromExpr(stayFoolish.expr());
                yield new SFMarkFiroe(stayFoolish, innerFir);
            }
            case AST.StayFullyFoolishExpr stayFullyFoolish ->
                new NKFiroe("SFF marker (<<==>> syntax) not yet implemented");
            default -> new NKFiroe();
        };
    }
}
